src/table_properties_rc.rs

```diff
@@ -0,0 +1,201 @@
+                }
+            }
+        }
+    }
+}
+
+/// # Safety
+///
+/// The underlying iterator is over an unordered map of heap-allocated strings,
+/// so as long as the iterator and collection are alive, the key pointers are
+/// valid.
+pub struct TablePropertiesKey {
+    key: *const u8,
+    keylen: size_t,
+    _iter_handle: TablePropertiesCollectionIteratorHandle,
+}
+
+impl TablePropertiesKey {
+    fn new(
+        key: *const u8,
+        keylen: size_t,
+        _iter_handle: TablePropertiesCollectionIteratorHandle,
+    ) -> TablePropertiesKey {
+        TablePropertiesKey {
+            key,
+            keylen,
+            _iter_handle,
+        }
+    }
+}
+
+impl Deref for TablePropertiesKey {
+    type Target = str;
+
+    fn deref(&self) -> &str {
+        // Safety: creating slice from values reported by rocksdb, that should
+        // be valid as long is this object is valid.
+        unsafe {
+            let bytes = slice::from_raw_parts(self.key, self.keylen);
+            str::from_utf8(bytes).unwrap()
+        }
+    }
+}
+
+pub struct TableProperties {
+    handle: TablePropertiesHandle,
+}
+
+impl TableProperties {
+    fn new(handle: TablePropertiesHandle) -> TableProperties {
+        TableProperties { handle }
+    }
+
+    fn get_u64(&self, prop: DBTableProperty) -> u64 {
+        unsafe { crocksdb_ffi::crocksdb_table_properties_get_u64(self.handle.ptr(), prop) }
+    }
+
+    pub fn num_entries(&self) -> u64 {
+        self.get_u64(DBTableProperty::NumEntries)
+    }
+
+    pub fn user_collected_properties(&self) -> UserCollectedProperties {
+        UserCollectedProperties::new(self.handle.clone())
+    }
+}
+
+pub struct UserCollectedProperties {
+    handle: UserCollectedPropertiesHandle,
+}
+
+impl UserCollectedProperties {
+    fn new(table_props_handle: TablePropertiesHandle) -> UserCollectedProperties {
+        UserCollectedProperties {
+            handle: UserCollectedPropertiesHandle::new(table_props_handle),
+        }
+    }
+
+    pub fn iter(&self) -> UserCollectedPropertiesIter {
+        UserCollectedPropertiesIter::new(self.handle.clone())
+    }
+
+    pub fn get<Q: AsRef<[u8]>>(&self, index: Q) -> Option<&[u8]> {
+        let bytes = index.as_ref();
+        let mut size = 0;
+        unsafe {
+            let ptr = crocksdb_ffi::crocksdb_user_collected_properties_get(
+                self.handle.ptr(),
+                bytes.as_ptr(),
+                bytes.len(),
+                &mut size,
+            );
+            if ptr.is_null() {
+                return None;
+            }
+            Some(slice::from_raw_parts(ptr, size))
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        unsafe { crocksdb_ffi::crocksdb_user_collected_properties_len(self.handle.ptr()) }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+pub struct UserCollectedPropertiesIter {
+    _handle: UserCollectedPropertiesIteratorHandle,
+}
```

Comment:
There has not been an implement of `Iteraor` for `UserCollectedPropertiesIter `, has there?

Reply:
No it isn't implemented, and not needed by TiKV. I've deleted the type. Thanks.